cognite/client/_api/spaces.py

```diff
@@ -0,0 +1,190 @@
+        return self._retrieve_multiple(list_cls=SpaceList, resource_cls=Space, identifiers=identifiers)
+
+    def list(
+        self,
+        partitions: int = None,
+        limit: int = LIST_LIMIT_DEFAULT,
+    ) -> SpaceList:
+        """`List events <https://docs.cognite.com/api/v1/#tag/Spaces/operation/listSpacesV3>`_
+
+        Args:
+            partitions (int): Retrieve events in parallel using this number of workers. Also requires `limit=None` to be passed.
+            limit (int, optional): Maximum number of events to return. Defaults to 25. Set to -1, float("inf") or None
+                to return all items.
+
+        Returns:
+            SpaceList: List of requested events
+
+        Examples:
+
+            List events and filter on max start time::
+
+                >>> from cognite.client import CogniteClient
+                >>> c = CogniteClient()
+                >>> event_list = c.models.spaces.list(limit=5)
+
+            Iterate over events::
+
+                >>> from cognite.client import CogniteClient
+                >>> c = CogniteClient()
+                >>> for space in c.models.spaces:
+                ...     space # do something with the space
+
+            Iterate over chunks of events to reduce memory load::
+
+                >>> from cognite.client import CogniteClient
+                >>> c = CogniteClient()
+                >>> for space_list in c.events(chunk_size=2500):
+                ...     space_list # do something with the spaces
+        """
+        return self._list(
+            list_cls=SpaceList,
+            resource_cls=Space,
+            method="GET",
+            limit=limit,
+            partitions=partitions,
+        )
+
+    @overload
+    def apply(self, space: Sequence[Space]) -> SpaceList:
+        ...
+
+    @overload
+    def apply(self, space: Space) -> Space:
+        ...
+
+    def apply(self, space: Space | Sequence[Space]) -> Space | SpaceList:
+        """`Create or update one or more spaces. <https://docs.cognite.com/api/v1/#tag/Spaces/operation/ApplySpaces>`_
+
+        Args:
+            space (space: Space | Sequence[Space]): Space or spaces of events to create or update.
+
+        Returns:
+            Space | SpaceList: Created space(s)
+
+        Examples:
+
+            Create new events::
+
+                >>> from cognite.client import CogniteClient
+                >>> from cognite.client.data_classes import Space
+                >>> c = CogniteClient()
+                >>> spaces = [Space(space="mySpace", description="My first space", name="My Space"),
+                ... Space(space="myOtherSpace", description="My second space", name="My Other Space")]
+                >>> res = c.models.spaces.create(spaces)
+        """
+        return self._create_multiple(list_cls=SpaceList, resource_cls=Space, items=space)
+
+    def delete(
+        self,
+        space: str | Sequence[str],
+    ) -> None:
```

Comment:
I think this needs to return the deleted spaces. Since the DM API doesn't support `ignoreUnknownIds`, the only the way the user can now if their space was successfully deleted or not is if it's present in the result.

Reply:
Agreed, I set it up to return a list of Spaces. 

This required some changes in the `_delete_multiple `in the `APIClient`, which again led me to change `geospatial` and `templates` as these were returning from their `_delete_multiple` call even though previously this would always be None. I removed the return to preserve the current behavior.

Reply:
May I ask why? 

`raw.databasesdelete(..)` returns => None, and has no ignore_unkown_ids
- if it returns None, everything was deleted successfully.
- if there is one space, which cannot be delete (doesn't exist, or contains elements) it throws an exception AKAIK

Reply:
FYI: [slack thread](https://cognitedata.slack.com/archives/C031G8Y19HP/p1675166822258489?thread_ts=1675115136.222279&cid=C031G8Y19HP) about improving the error msg in this case

Reply:
Thanks for this additional information. I think that should be improved for the `raw.databases.delete(...)` rather than copy that behavior her. When it comes to the exception if there are nodes and edges, I suggest to keep it as is for now, and when I add support for the rest I will add testing for it as well. And hopefully, soon, the error message from CDF will improve so then we can update the SDK to take advantage of this as well.

Reply:
Good comment to better fix raw delete!
Fine for me to proceed